src/lava/magma/core/model/py/ports.py

```diff
@@ -9,96 +9,229 @@
+    which is a list of CSP ports corresponding to the connected PyPorts.
+    """
     @property
     @abstractmethod
     def csp_ports(self) -> ty.List[AbstractCspPort]:
-        """Returns all csp ports of the port."""
+        """
+        Abstract property to get the corresponding CSP Ports of all connected
+        PyPorts (csp_ports). The CSP Port is the low level interface of the
+        backend messaging infrastructure which is used to send and receive data.
+
+        Returns
+        -------
+        A list of all CSP Ports connected to the PyPort.
+        """
         pass
 
 
-class PyInPort(AbstractPyPort):
+class AbstractPyIOPort(AbstractPyPort):
+    """Abstract class of an input/output Port implemented in python.
+
+    A PyIOPort can either be an input or an output Port and is the common
+    abstraction of PyInPort/PyOutPort.
+    _csp_ports is a list of CSP Ports which are used to send/receive data by
+    connected PyIOPorts.
+
+    Parameters
+    ----------
+    csp_ports : list
+        A list of csp Ports used by this IO Port.
+
+    process_model : AbstractProcessModel
+        The process model used by the process of the Port.
+
+    shape : tuple
+        The shape of the Port.
+
+    d_type: type
+        The data type of the Port.
+
+    Attributes
+    ----------
+    _csp_ports : list
+        A list of CSP Ports used by this IO Port.
+    """
+    def __init__(self,
+                 csp_ports: ty.List[AbstractCspPort],
+                 process_model: AbstractProcessModel,
+                 shape: ty.Tuple[int, ...],
+                 d_type: type):
+
+        self._csp_ports = csp_ports
+        super().__init__(process_model, shape, d_type)
+
+    @property
+    def csp_ports(self) -> ty.List[AbstractCspPort]:
+        """Property to get the corresponding CSP Ports of all connected
+        PyPorts (csp_ports). The CSP Port is the low level interface of the
+        backend messaging infrastructure which is used to send and receive data.
+
+        Returns
+        -------
+        A list of all CSP Ports connected to the PyPort.
+        """
+        return self._csp_ports
+
+
+class PyInPort(AbstractPyIOPort):
     """Python implementation of InPort used within AbstractPyProcessModel.
-    If buffer is empty, recv() will be blocking.
+
+    PyInPort is an input Port processing data sent from a connected output Port
+    (PyOutPort) over a channel. PyInPort can receive (recv()) the data, which
+    removes it from the channel, look at (peek()) at the data which keeps it on
+    the channel or check (probe()) if there is data on the channel.
+    The different class attributes are used to select the type of OutPorts via
+    LavaPyType declarations in PyProcModels, e.g., LavaPyType(
+    PyInPort.VEC_DENSE, np.int32, precision=24) creates a PyInPort.
+    A PyOutPort (source) can be connected to one or multiple PyInPorts (target).
+
```

Comment:
May not be fully clear what "processing data" means.
What's about: "PyInPort is an input Port that can be instantiated by a Process to receive data sent from a connected ->PyOutPort of another Process over a channel."

"look ~~at~~ (peek()) at the data"

Unclear why it is import to remove data from a channel, should be explained in a sentence.

"A PyOutPort (source) can be connected to one or multiple PyInPorts (target)."
This docstring is about PyInPorts. Thus, I would here turn it around so that PyInPorts are the important aspects of this sentence:
"Multiple PyInPorts (target) can be connected to the same PyOutPort (source).

Reply:
Applied some suggestions